src/Router.ts

```diff
@@ -172,12 +172,89 @@ export class Router<C extends Context> extends Evented implements RouterInterfac
 		}
 	}
 
-	dispatch(context: Context, path: string): Task<DispatchResult> {
-		// Reset, any further calls can't have come from start(). This is necessary since the navstart listeners
-		// may call dispatch() themselves.
-		let dispatchFromStart = this._dispatchFromStart;
+	private _dispatch(context: Context, path: string, canceled: boolean = false, emit: boolean = true) {
+		if (canceled) {
+			return { success: false };
+		}
+
+		this._currentParams = {};
+		this._outletContextMap.clear();
+
+		if (emit) {
+			this.emit<NavigationStartEvent>({
+				context,
+				path,
+				target: this,
+				type: 'navstart'
+			});
+		}
+
+		const { searchParams, segments, trailingSlash } = parsePath(path);
+		const dispatchFromStart = this._dispatchFromStart;
 		this._dispatchFromStart = false;
 
+		let redirect: undefined | string;
+		const dispatched = this._routes.some((route) => {
+			const result = route.select(context, segments, trailingSlash, searchParams);
+
+			if (typeof result === 'string') {
+				redirect = result;
+				return true;
+			}
+			if (result.length === 0) {
+				return false;
+			}
+
+			// Update the selected routes after selecting new routes, but before invoking the handlers.
+			// This means the original value is available to guard() and params() functions, and the
+			// new value when the newly selected routes are executed.
+			//
+			// Reset selected routes if not dispatched from start().
+			this._currentSelection = dispatchFromStart ? result : [];
+
+			for (const { handler, params, outlet, type, route } of result) {
+				if (outlet) {
+					if (params) {
```

Comment:
We don't seem to have a !params scenario... is it really just that params will always exist and isn't optional?

Reply:
This one, I don't think was covered originally (highlighted because I moved the codes). I'll check `params` and see if it's always at least an empty object.